xdsl/dialects/csl.py

```diff
@@ -757,6 +757,119 @@ def verify_(self) -> None:
             )
 
 
+@irdl_op_definition
+class SetDsdBaseAddrOp(IRDLOperation):
+    """
+    Returns a clone of the DSD with a different base_addr.
+    Only works on memory DSDs, i.e. mem1d_dsd or mem4d_dsd.
+
+    Implements the CSL built-in
+    @set_dsd_base_addr(input_dsd, base_addr)
+    """
+
+    name = "csl.set_dsd_base_addr"
+
+    op = operand_def(DsdType)
+    base_addr = operand_def(MemRefType | PtrType)
+    result = result_def(DsdType)
+
+    def verify_(self) -> None:
+        if (
+            not isinstance(self.result.type, DsdType)
+            or not isinstance(self.op.type, DsdType)
+            or self.result.type.data not in [DsdKind.mem1d_dsd, DsdKind.mem4d_dsd]
+            or self.op.type.data not in [DsdKind.mem1d_dsd, DsdKind.mem4d_dsd]
+        ):
+            raise VerifyException(f"{self.name} must operate on mem1d_dsd or mem4d_dsd")
+        if (
+            isinstance(self.base_addr.type, PtrType)
+            and not self.base_addr.type.kind.data == PtrKind.MANY
+        ):
+            raise VerifyException(
+                f"{self.name} cannot operate on pointers of kind {self.base_addr.type}"
+            )
+
+
+@irdl_op_definition
+class IncrementDsdOffsetOp(IRDLOperation):
+    """
+    Returns a clone of the DSD with a different offset
+    Only works on memory DSDs, i.e. mem1d_dsd or mem4d_dsd.
+
+    Implements the CSL built-in
+    @increment_dsd_offset(input_dsd, offset, elem_type)
+
+    where offset is a 16-bit signed int that may be negative,
+    and elem_type is used to convert offset into words (any u,i,f type of 16,32 bit)
+    elem_type should be derived by the printer
```

Comment:
Since the DSD type doesn't carry its element type, this would be quite complicated. It would make things much more straightforward if it was passed here as a property.

Alternatively, DSD type could carry this information.

Reply:
Rewrote `DsdType` to carry element type. Can be set to `none` for fabric DSDs